### PR TITLE
ENH: stats.permutation_test: improve performance of samples/pairings random permutations

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-16962

#### What does this implement/fix?
gh-16962 brought up a case in which `permutation_test` was unreasonably slow. This is because `RandomState.permutation` doesn't operate on each axis independently, so we loop in pure Python to generate arrays of permutations. We can do better.

- Way back when I first ran into this, I found that `np.argsort`ing a `random` array tends to be a much faster way of generating the large arrays of random permutations we need. I was [advised against using this approach](https://github.com/scipy/scipy/pull/4824#discussion_r529271983). I agree with the sentiment to some extent, but all things considered, I have to disagree here because the "hackery" can be _so_ much better. If someone finds a case in which the hack is less performant in the future, they should just use `Generator`. I don't think we should need to Cythonize this.
- `Generator` has a `permuted` method that _does_ act on slices independently, so use that when it's available.

This PR uses these faster methods of generating arrays of permutations to improve performance substantially (25~50x).

```python3
import numpy as np
from scipy import stats

rng = np.random.default_rng()
rvs = rng.normal(size=1000)
def t_statistic(x, axis=-1):
    return stats.ttest_1samp(x, popmean=0, axis=axis).statistic

res = stats.permutation_test((rvs,), t_statistic, permutation_type='samples')
# PR: 671 ms ± 22.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# main: 17.9 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
res2 = stats.permutation_test((rvs,), t_statistic, permutation_type='samples',
                              random_state=rng)
# PR: 561 ms ± 11.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# main: 27.7 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
```

#### Additional information
- I might do a follow-up that implements this for the `independent` permutation type.
- I've written some (batched) Pythran exhaustive permutation generators for the exact tests. Eventually, we'll want to replace the pure-Python loops over `itertools.combinations`/`itertools.permutations` with those. Like `permuted`, these could permute the samples themselves rather than permuting indices and then doing fancy indexing. It would also avoid all the set difference stuff in `all_partitions`. It would save memory and improve speed if we took advantage of that.
- Eventually, we could improve  `_calculate_null_samples` by writing a standalone implementation (without having it call `_calculate_null_pairings`), but I found that didn't improve things very much unless `batch` is small and the number of observations per sample is quite large. Let's do this in chunks to keep reviews manageable.

### CODE REVIEW

File: scipy/stats/_resampling.py [hunk 1]
```diff
@@ -893,6 +893,30 @@ def _batch_generator(iterable, batch):
         z = [item for i, item in zip(range(batch), iterator)]
 
 
+def _pairings_permutations_gen(n_permutations, n_samples, n_obs_sample, batch,
+                               random_state):
+
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:898]:
```suggestion
    # This function returns a generator that yields batches of permuted indices
    # of shape (batch, n_samples, n_obs_sample)
```

File: scipy/stats/_resampling.py [hunk 2]
```diff
@@ -893,6 +893,30 @@ def _batch_generator(iterable, batch):
...
+    if isinstance(random_state, np.random.Generator):
+        def batched_perm_generator():
+            indices = np.arange(n_obs_sample)
+            indices = np.tile(indices, (batch, n_samples, 1))
+            for k in range(0, n_permutations, batch):
+                batch_actual = min(batch, n_permutations-k)
+                # Don't permute in place, otherwise results depend on `batch`
+                permuted_indices = random_state.permuted(indices, axis=-1)
+                yield permuted_indices[:batch_actual]
+    else:
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:910]:
```suggestion
    else:  # RandomState doesn't have `permuted`
```

File: scipy/stats/_resampling.py [hunk 3]
```diff
@@ -969,22 +995,23 @@ def _calculate_null_pairings(data, statistic, n_permutations, batch,
     if n_permutations >= n_max:
         exact_test = True
         n_permutations = n_max
+        batch = batch or int(n_permutations)
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:998]: The default of `batch` is `None`.

File: scipy/stats/_resampling.py [hunk 4]
```diff
@@ -969,22 +995,23 @@ def _calculate_null_pairings(data, statistic, n_permutations, batch,
...
         # cartesian product of the sets of all permutations of indices
         perm_generator = product(*(permutations(range(n_obs_sample))
                                    for i in range(n_samples)))
+        batched_perm_generator = _batch_generator(perm_generator, batch=batch)
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:1002]: In this case - exhaustive test of all permutations - we're not changing any behavior. These changed lines have essentially been moved up from below.

File: scipy/stats/_resampling.py [hunk 5]
```diff
@@ -969,22 +995,23 @@ def _calculate_null_pairings(data, statistic, n_permutations, batch,
...
         exact_test = False
+        batch = batch or int(n_permutations)
         # Separate random permutations of indices for each sample.
         # Again, it would be nice if RandomState/Generator.permutation
         # could permute each axis-slice separately.
-        perm_generator = ([random_state.permutation(n_obs_sample)
-                           for i in range(n_samples)]
-                          for j in range(n_permutations))
+        args = n_permutations, n_samples, n_obs_sample, batch, random_state
+        batched_perm_generator = _pairings_permutations_gen(*args)
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:1010]: This is what we're changing. That looping was really slow.

File: scipy/stats/tests/test_resampling.py [hunk 6]
```diff
@@ -940,9 +940,11 @@ def stat(x, y, axis):
                              random_state='herring')
 
     # -- Test Parameters -- #
+    @pytest.mark.parametrize('random_state', [np.random.RandomState,
+                                              np.random.default_rng])
```
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:944]: It's important to test with both `RandomState` and `default_rng` because we use them in different ways.

**Review by mdhaber — COMMENTED** [review]
@tupui I saw the rocket emoji. Does that mean you're willing to take a look?

**Review by tupui — COMMENTED** [review]
I rocketed after having a quick look. I am +1 with your approach. I did not review the generator logic itself, but tests are passing and performance greatly improved so I would be inclined to trust your code. I could do more if you wish.

Just some nitpicking/questions otherwise.

File: scipy/stats/tests/test_resampling.py [hunk 7]
```diff
@@ -958,29 +960,34 @@ def statistic(x, y, axis):
...
     @pytest.mark.parametrize('permutation_type, exact_size',
                              [('pairings', special.factorial(3)**2),
                               ('samples', 2**3),
                               ('independent', special.binom(6, 3))])
-    def test_permutations(self, permutation_type, exact_size):
+    def test_permutations(self, permutation_type, exact_size, random_state):
         # make sure that the `permutations` parameter is respected by checking
         # the size of the null distribution
         np.random.seed(0)
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/tests/test_resampling.py]: Just for you, I did: 5180e9f48104a50cedbc32f817ee850c34b69d85

Any uses of `RandomState` left in `TestPermutationTest` (very few) are intentional.

**Comment by mdhaber** [comment]
> I could do more if you wish.

I'd appreciate it if you'd take a closer look. Maybe just confirming my timing comparison locally, reading through the changes enough to see how they work, and seeing whether you agree it is OK that this PR reverses my decision to take [this advice](https://github.com/scipy/scipy/pull/4824#discussion_r529271983). Thanks!

**Review by tupui — APPROVED** [review]
I had another look and could also reproduce your timings. LGTM and I think we should move forward with it.

**Comment by mdhaber** [comment]
Thanks @tupui!

**Pull request merged**